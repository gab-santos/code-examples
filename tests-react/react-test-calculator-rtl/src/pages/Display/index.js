import React from "react";

import "./styles.css";

export default function Display({ displayValue }) {
  return (
    <div testid="display-container" className="display-container">
      <p testid="display-value" className="display-value">
        {displayValue}
      </p>
    </div>
  );
}
