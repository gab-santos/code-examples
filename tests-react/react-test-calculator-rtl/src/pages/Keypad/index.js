import React from "react";

// import { Container } from './styles';

export default function Keypad() {
  return <div testid="keypad-container" className="keypad-container" />;
}
