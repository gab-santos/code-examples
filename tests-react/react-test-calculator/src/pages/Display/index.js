import React from "react";

import "./styles.css";

export default function Display({ displayValue, setDisplayValue }) {
  return (
    <div className="display-container">
      <p className="display-value">{displayValue}</p>
    </div>
  );
}
