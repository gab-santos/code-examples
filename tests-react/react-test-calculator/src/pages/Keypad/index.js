import React from "react";

import "./styles.css";
import Key from "../Key";

export default function Keypad({
  numbers,
  operators,
  callOperator,
  setOperator,
  updateDisplay
}) {
  const numberKeys = numbers.map(number => (
    <Key
      key={number}
      keyAction={updateDisplay}
      keyType="number-key"
      keyValue={number}
    />
  ));

  const operatorKeys = operators.map(operator => (
    <Key
      key={operator}
      keyAction={setOperator}
      keyType="operator-key"
      keyValue={operator}
    />
  ));

  return (
    <div className="keypad-container">
      <div className="numbers-container">{numberKeys}</div>
      <div className="operators-container">{operatorKeys}</div>
      <div className="submit-container">
        <Key keyAction={callOperator} keyType="submit-key" keyValue="=" />
      </div>
    </div>
  );
}
