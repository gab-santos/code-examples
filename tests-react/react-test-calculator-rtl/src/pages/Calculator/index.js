import React, { useState } from "react";

import "./styles.css";
import Display from "../Display";
import Keypad from "../Keypad";

export default function Calculator() {
  // value to be displayed in <Display />
  const [displayValue, setDisplayValue] = useState("0");

  // value to be displayed in <Keys />
  const [numbers, setNumbers] = useState([]);

  // value to be displayed in operator <Keys />
  const [operators, setOperators] = useState([]);

  // operator selected for math operation
  const [selectedOperator, setSelectedOperator] = useState("");

  // stored value to use for math operation
  const [storedValue, setStoredValue] = useState("");

  function updateDisplay() {
    console.log("Update display");
  }

  function callOperator() {
    console.log("Call operation");
  }

  function setOperator() {
    console.log("Set operation");
  }

  return (
    <div testid="calculator-container" className="calculator-container">
      <Display displayValue={displayValue} />
      <Keypad numbers={numbers} operators={operators} />
    </div>
  );
}
