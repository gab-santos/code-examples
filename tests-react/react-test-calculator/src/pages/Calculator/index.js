import React, { useState } from "react";

import "./styles.css";
import Display from "../Display";
import Keypad from "../Keypad";

export default function Calculator() {
  // value to be displayed in <Display />
  let [displayValue, setDisplayValue] = useState("0");

  // value to be displayed in <Keys />
  const [numbers, setNumbers] = useState([
    "9",
    "8",
    "7",
    "6",
    "5",
    "4",
    "3",
    "2",
    "1",
    ".",
    "0",
    "ce"
  ]);

  // value to be displayed in operator <Keys />
  const [operators, setOperators] = useState(["/", "*", "-", "+"]);

  // operator selected for math operation
  const [selectedOperator, setSelectedOperator] = useState("");

  // stored value to use for math operation
  const [storedValue, setStoredValue] = useState("");

  function updateDisplay(value) {
    // prevet multiple ocorrences of "."
    if (value === "." && displayValue.includes(".")) value = "";

    if (value === "ce") {
      // delete last char in displayValue
      setDisplayValue(displayValue.substr(0, displayValue.length - 1));
      // set display value to '0' if displayValue is empty string
      if (displayValue === "0") setDisplayValue("0");
    } else {
      displayValue === "0"
        ? // replace displayValue with value if display value equal to '0'
          // else concatenate displayValue and value
          setDisplayValue(value)
        : setDisplayValue((displayValue += value));
    }
  }

  function callOperator() {
    console.log("Call operation");
  }

  function setOperator() {
    console.log("Set operation");
  }

  return (
    <div className="calculator-container">
      <Display displayValue={displayValue} setDisplayValue={setDisplayValue} />
      <Keypad
        numbers={numbers}
        setNumbers={setNumbers}
        operators={operators}
        setOperators={setOperators}
        callOperator={callOperator}
        setOperator={setOperator}
        updateDisplay={updateDisplay}
      />
    </div>
  );
}
