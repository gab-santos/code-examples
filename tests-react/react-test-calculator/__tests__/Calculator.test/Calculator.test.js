import React from "react";
import { shallow, mount } from "enzyme";

import Calculator from "../../src/pages/Calculator";
import Display from "../../src/pages/Display";
import Keypad from "../../src/pages/Keypad";

describe("Calculator", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Calculator />);
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a <div>", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });

  it("should render the Display and Keypad Component", () => {
    expect(
      wrapper.containsAllMatchingElements([
        <Display
          displayValue={wrapper.props().displayValue}
          setDisplayValue={wrapper.props().setDisplayValue}
        />,
        <Keypad
          //
          numbers={wrapper.props().numbers}
          setNumbers={wrapper.props().setNumbers}
          //
          operators={wrapper.props().operators}
          setOperators={wrapper.props().setOperators}
          //
          callOperator={wrapper.props().callOperator}
          setOperator={wrapper.props().setOperator}
          updateDisplay={wrapper.props().updateDisplay}
        />
      ])
    ).toEqual(true);
  });
});

describe("updateDisplay", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Calculator />);
  });

  it("should updates displayValue", () => {
    wrapper
      .find(".number-key")
      .at(4)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("5");
  });

  it("should concatenates displayValue", () => {
    wrapper
      .find(".number-key")
      .at(4)
      .simulate("click");
    wrapper
      .find(".number-key")
      .at(10)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("50");
  });

  it('should removes loading "0" from displayValue', () => {
    wrapper
      .find(".number-key")
      .at(10)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("0");

    wrapper.update();
    wrapper
      .find(".number-key")
      .at(4)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("5");
  });

  it('should prevents multiple instances of "." in display value', () => {
    wrapper
      .find(".number-key")
      .at(9)
      .simulate("click");
    wrapper
      .find(".number-key")
      .at(9)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual(".");
  });

  it("should removes last char of displayValue", () => {
    wrapper
      .find(".number-key")
      .at(4)
      .simulate("click");
    wrapper
      .find(".number-key")
      .at(10)
      .simulate("click");
    wrapper
      .find(".number-key")
      .at(11)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("5");
  });

  it('should will set displayValue to "0" if displayValue is equal to a empty string', () => {
    wrapper
      .find(".number-key")
      .at(11)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("0");
  });

  it('should prevents multiple leaging "0"s from displayValue', () => {
    wrapper
      .find(".number-key")
      .at(10)
      .simulate("click");
    wrapper
      .find(".number-key")
      .at(10)
      .simulate("click");

    expect(wrapper.find(".display-value").text()).toEqual("0");
  });
});

// Mounted Calculator
{
  // describe("Mounted calculator", () => {
  // let wrapper;
  // beforeEach(() => {
  //   wrapper = mount(<Calculator />);
  // });
  // it("should calls updateDisplay when a number key is clicked", () => {
  //   wrapper.updateDisplay = jest.fn();
  //   const spy = jest.spyOn(wrapper, "updateDisplay");
  //   wrapper.update();
  //   wrapper
  //     .find(".number-key")
  //     .first()
  //     .simulate("click");
  //   expect(spy).toHaveBeenCalledTimes(1);
  // });
  // it("should calls setOperator when a operator key is clicked", () => {
  //   wrapper.setOperator = jest.fn();
  //   const spy = jest.spyOn(wrapper, "setOperator");
  //   wrapper.update();
  //   wrapper
  //     .find(".operator-key")
  //     .first()
  //     .simulate("click");
  //   expect(spy).toHaveBeenCalledTimes(1);
  // });
  // it("should calls callOperator when the submit key is clicked", () => {
  //   wrapper.callOperator = jest.fn();
  //   const spy = jest.spyOn(wrapper, "callOperator");
  //   wrapper.update();
  //   wrapper
  //     .find(".submit-key")
  //     .first()
  //     .simulate("click");
  //   expect(spy).toHaveBeenCalledTimes(1);
  // });
  // });
}
