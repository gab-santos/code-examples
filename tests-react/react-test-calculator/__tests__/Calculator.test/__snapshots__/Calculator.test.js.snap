// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render correctly 1`] = `
<div
  className="calculator-container"
>
  <Display
    displayValue="0"
    setDisplayValue={[Function]}
  />
  <Keypad
    callOperator={[Function]}
    numbers={
      Array [
        "9",
        "8",
        "7",
        "6",
        "5",
        "4",
        "3",
        "2",
        "1",
        ".",
        "0",
        "ce",
      ]
    }
    operators={
      Array [
        "/",
        "*",
        "-",
        "+",
      ]
    }
    setNumbers={[Function]}
    setOperator={[Function]}
    setOperators={[Function]}
    updateDisplay={[Function]}
  />
</div>
`;
